File: julia/day11_idiomatic.jl
# Day 11: Graph Path Counting with Memoized DFS
# Idiomatic Julia version using adjacency lists for better performance

include("aoc_common_idiomatic.jl")

function parse_line(line::AbstractString)::Tuple{String, Vector{String}}
    """Parse line into source and destinations."""
    parts = split(line, ':')
    return (String(parts[1]), String.(split(strip(parts[2]))))
end

function make_node_table(input_str::String)::Dict{String, Int}
    """Create mapping from node names to IDs (0-indexed for consistency)."""
    table = Dict{String, Int}()
    next_id = 0
    for line in split(strip(input_str), '\n')
        isempty(line) && continue
        src, dsts = parse_line(line)
        if !haskey(table, src)
            table[src] = next_id
            next_id += 1
        end
        for dst in dsts
            if !haskey(table, dst)
                table[dst] = next_id
                next_id += 1
            end
        end
    end
    return table
end

function build_adjacency_list(input_str::String, table::Dict{String, Int})::Vector{Vector{Int}}
    """Build adjacency list representation (much more efficient than matrix for sparse graphs)."""
    n = length(table)
    adj = [Int[] for _ in 1:n]

    for line in split(strip(input_str), '\n')
        isempty(line) && continue
        src, dsts = parse_line(line)
        src_id = table[src]
        for dst in dsts
            dst_id = table[dst]
            push!(adj[src_id + 1], dst_id)  # +1 for Julia 1-indexing
        end
    end

    return adj
end

function process_input(input_str::String)::Tuple{Dict{String, Int}, Vector{Vector{Int}}}
    """Parse input into node table and adjacency list."""
    table = make_node_table(input_str)
    adj = build_adjacency_list(input_str, table)
    return table, adj
end

# Part 1: Simple DFS to count all paths (much faster with adjacency lists!)
function dfs(from_node::Int, to_node::Int, adj::Vector{Vector{Int}})::Int
    """Count all paths from from_node to to_node."""
    from_node == to_node && return 1

    # Sum paths from all children (only iterate over actual neighbors!)
    return sum(dfs(child, to_node, adj) for child in adj[from_node + 1]; init=0)
end

expect("day11_part1", "5")
function day11_part1(input_str::String)::String
    table, adj = process_input(input_str)
    you = table["you"]
    out = table["out"]
    return string(dfs(you, out, adj))
end

# Part 2: DFS with memoization
function dfs_p2(from_node::Int, to_node::Int, dac_id::Int, fft_id::Int, adj::Vector{Vector{Int}})::Int
    """Count paths that visit both dac and fft nodes."""
    memo = Dict{Tuple{Int, Bool, Bool}, Int}()

    function dfs_inner(node::Int, seen_dac::Bool, seen_fft::Bool)::Int
        # Update seen flags
        seen_dac |= (node == dac_id)
        seen_fft |= (node == fft_id)

        # Check memo
        key = (node, seen_dac, seen_fft)
        haskey(memo, key) && return memo[key]

        # Base case
        if node == to_node
            result = (seen_dac && seen_fft) ? 1 : 0
        else
            # Recursive case: sum paths from children
            result = sum(dfs_inner(child, seen_dac, seen_fft) for child in adj[node + 1]; init=0)
        end

        return memo[key] = result
    end

    return dfs_inner(from_node, false, false)
end

expect("day11_part2", "2")
function day11_part2(input_str::String)::String
    table, adj = process_input(input_str)
    svr, out = table["svr"], table["out"]
    dac, fft = table["dac"], table["fft"]
    return string(dfs_p2(svr, out, dac, fft, adj))
end

# Main execution
if abspath(PROGRAM_FILE) == @__FILE__
    # Run on real input (no sample file for day11)
    real = read_input(11)
    println("Running Part 1 on real input:")
    println(day11_part1(real))

    println("\nRunning Part 2 on real input:")
    println(day11_part2(real))
end

File: python/day11_idiomatic.py
"""
Day 11: Graph Path Counting with Memoized DFS
Idiomatic Python version using NumPy and @lru_cache.
"""
import sys
sys.path.insert(0, '/Users/dpzmick/programming/elisp-aoc/python')

from functools import lru_cache
import numpy as np
from aoc_common_idiomatic import read_input, read_sample, expect, check_result


def parse_line(line: str) -> tuple[str, list[str]]:
    """Parse line into source and destinations."""
    parts = line.split(':')
    src = parts[0]
    dsts = parts[1].strip().split()
    return src, dsts


def make_node_table(input_str: str) -> dict[str, int]:
    """Create mapping from node names to IDs."""
    nodes = set()
    for line in input_str.strip().split('\n'):
        if not line:
            continue
        src, dsts = parse_line(line)
        nodes.add(src)
        nodes.update(dsts)

    return {name: i for i, name in enumerate(sorted(nodes))}


def process_input(input_str: str) -> tuple[dict[str, int], np.ndarray]:
    """Parse input into node table and adjacency matrix."""
    table = make_node_table(input_str)
    n = len(table)
    adj = np.zeros((n, n), dtype=np.int32)

    for line in input_str.strip().split('\n'):
        if not line:
            continue
        src, dsts = parse_line(line)
        row = table[src]
        for dst in dsts:
            col = table[dst]
            adj[row, col] = 1

    return table, adj


# Part 1: Simple DFS to count all paths
def dfs(from_node: int, to_node: int, adj: np.ndarray) -> int:
    """Count all paths from from_node to to_node."""
    if from_node == to_node:
        return 1

    # Sum paths from all children
    children = np.where(adj[from_node] == 1)[0]
    return sum(dfs(child, to_node, adj) for child in children)


expect('day11_part1', '5')
def day11_part1(input_str: str) -> str:
    table, adj = process_input(input_str)
    you = table['you']
    out = table['out']
    return str(dfs(you, out, adj))


# Part 2: DFS with memoization tracking specific nodes visited
def dfs_p2(from_node: int, to_node: int, dac_id: int, fft_id: int, adj: np.ndarray) -> int:
    """Count paths that visit both dac and fft nodes."""

    @lru_cache(maxsize=None)
    def dfs_inner(from_node: int, seen_dac: bool, seen_fft: bool) -> int:
        # Update seen flags
        seen_dac = seen_dac or (from_node == dac_id)
        seen_fft = seen_fft or (from_node == fft_id)

        # Base case
        if from_node == to_node:
            return 1 if (seen_dac and seen_fft) else 0

        # Recursive case: sum paths from children
        children = np.where(adj[from_node] == 1)[0]
        return sum(dfs_inner(child, seen_dac, seen_fft) for child in children)

    return dfs_inner(from_node, False, False)


expect('day11_part2', '2')
def day11_part2(input_str: str) -> str:
    table, adj = process_input(input_str)
    svr = table['svr']
    out = table['out']
    dac = table['dac']
    fft = table['fft']
    return str(dfs_p2(svr, out, dac, fft, adj))


if __name__ == '__main__':
    # Test on sample input (sample only has part 2 data)
    sample = read_sample(11)
    print("Testing Part 2 on sample:")
    result2 = day11_part2(sample)
    check_result('day11_part2', result2, 'sample')

    # Run on real input
    real = read_input(11)
    print("\nRunning Part 1 on real input:")
    print(day11_part1(real))

    print("\nRunning Part 2 on real input:")
    print(day11_part2(real))
